File: LeetCode/33_Search_in_Rotated_Sorted_Array.py
def search(nums, target):
    left = 0
    right = len(nums) - 1
    
    while left <= right:
        middle = (left + right)//2
        if target == nums[middle]:
            return middle
        # left sorted portion
        if nums[left] <= nums[middle]:
            if target > nums[middle] or target < nums[left]:
                left = middle + 1
            else:
                right = middle - 1
        # right sorted portion
        else:
            if target < nums[middle] or target > nums[right]:
                right = middle - 1
            else:
                left = middle + 1 
    return -1
        


# a = [4,5,6,7,0,1,2] 
# b = 0

# a = [4,5,6,7,0,1,2] 
# b = 3

a = [1]
b = 0

print(search(a,b))

def search(nums, target):
    left, right = 0 , len(nums) - 1
    
    while left <= right:
        mid = (left + right) // 2
        if target == nums[mid]:
            return mid
        if nums[left] <= nums[mid]:
            if target > nums[mid] or target < nums[left]:
                left = mid + 1
            else:
                right = mid - 1
        if nums[mid] <= nums[right]:
            if target < nums[mid] or target > nums[right]:
                right = mid - 1
            else:
                left = mid + 1
    return -1

File: LeetCode/33_Search_in_Rotated_Sorted_Array.js
var search = function(nums, target) {
    var l = 0, r = nums.length - 1;
    while (l <= r) {
      var mid = Math.floor((l+r)/2)
      if (target === nums[mid]) {
        return mid
      }
      if (nums[l] <= nums[mid]) {
        if (target > nums[mid] || target < nums[l]) {
          l = mid + 1
        } else {
          r = mid - 1
        }
      } else {
        if (target < nums[mid] || target > nums[r]) {
          r = mid - 1
        } else {
          l = mid + 1
        }
      }
    }
    return - 1
};
var nums = [4,5,6,7,0,1,2], target = 0
search(nums, target)